### PR TITLE
Expose recursive queries for entity types

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Similar to #1051, this implements recursive lookups for entity types

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202882583267702/1202962778211049/f) _(internal)_

## 🔍 What does this change?

- Expose recursive lookup in GraphQL

## 📜 Does this require a change to the docs?

The GraphQL documentation was adjusted

## 🐾 Next steps

- Implement the same for enties and entities: [Asana task](https://app.asana.com/0/1200211978612931/1202682556665585/f) (_internal_)
## 📹 Demo

```gql
query recursive_entity_types {
  getEntityType(entityTypeVersionedUri: "http://localhost:3000/@example/types/entity-type/organization/v/1") {
    entityTypeVersionedUri

    referencedDataTypes {
      dataTypeVersionedUri
    }

    referencedPropertyTypes {
      propertyTypeVersionedUri
    }
    
    referencedLinkTypes {
      linkTypeVersionedUri
    }
    
    referencedEntityTypes {
      entityTypeVersionedUri
    }
  }
}
```

```json
{
  "data": {
    "getEntityType": {
      "entityTypeVersionedUri": "http://localhost:3000/@example/types/entity-type/organization/v/1",
      "referencedDataTypes": [
        {
          "dataTypeVersionedUri": "https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1"
        }
      ],
      "referencedPropertyTypes": [
        {
          "propertyTypeVersionedUri": "http://localhost:3000/@example/types/property-type/account-id/v/1"
        },
        {
          "propertyTypeVersionedUri": "http://localhost:3000/@example/types/property-type/shortname/v/1"
        },
        {
          "propertyTypeVersionedUri": "http://localhost:3000/@example/types/property-type/organization-provided-info/v/1"
        },
        {
          "propertyTypeVersionedUri": "http://localhost:3000/@example/types/property-type/organization-size/v/1"
        },
        {
          "propertyTypeVersionedUri": "http://localhost:3000/@example/types/property-type/organization-name/v/1"
        }
      ],
      "referencedLinkTypes": [],
      "referencedEntityTypes": []
    }
  }
}
```

### CODE REVIEW

File: packages/hash/api/src/graphql/typeDefs/ontology/entity-type.typedef.ts [hunk 1]
```diff
@@ -4,7 +4,66 @@ export const entityTypeTypedef = gql`
...
+    """
+    accountId: ID!
+    """
+    The entity type
+    """
+    entityType: EntityType!
+    # INTERFACE FIELDS END #
+  }
+
+  type EntityTypeSubgraph implements PersistedEntityTypeInterface {
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/graphql/typeDefs/ontology/entity-type.typedef.ts:48]: We're using `EntityTypeRootedSubgraph` in the Rust code right?

Appreciate it's a bit wordy but I think `EntityTypeSubgraph` _could_ be a little misleading into thinking it's a subgraph with only entity types in it. Rather than one rooted at an entity type

Comment (TimDiekmann) [reply packages/hash/api/src/graphql/typeDefs/ontology/entity-type.typedef.ts:48]: > We're using EntityTypeRootedSubgraph in the Rust code right?

Yes, we do, I'll change it

**Review by benwerner01 — APPROVED** [review]
Great work Tim, this looks good and is working locally 👍

Have left two small nits, but these don't block merging.

File: packages/hash/api/src/graphql/resolvers/util.ts [hunk 2]
```diff
@@ -80,3 +80,23 @@ export const dataTypeQueryDepth = (info: GraphQLResolveInfo): number => {
...
+/**
+ * Returns the requested query depth for referenced link types
+ *
+ * @param info the 'info' fourth argument to all resolvers
+ * @returns the depth requested
+ */
+export const linkTypeQueryDepth = (info: GraphQLResolveInfo): number => {
+  return queryDepthByFieldName(info, "referencedLinkTypes");
+};
+
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/hash/api/src/graphql/resolvers/util.ts:93]: I'd prefer sticking with the functions, as the functions are called in quite a few places. Passing the names everywhere is verbose. Additionally, we can change the default values for some functions as needed (e.g. for entities we probably want to set it to `1` instead of `255`).

File: packages/hash/api/src/graphql/typeDefs/ontology/entity-type.typedef.ts (hunk 1 above)
Comment (TimDiekmann) [reply packages/hash/api/src/graphql/typeDefs/ontology/entity-type.typedef.ts]: Applied in https://github.com/hashintel/hash/pull/1062/commits/533f32d049be71f86a0978939c3a83add8ca6bc9

File: packages/hash/api/src/graphql/resolvers/util.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/util.ts:93]: > Additionally, we can change the default values for some functions as needed (e.g. for entities we probably want to set it to 1 instead of 255)

I hadn't considered that, makes sense 👍

**Review by Alfred-Mountfield — APPROVED** [review]
Re-approving for CI
